diag_manager/fms_diag_yaml.F90

```diff
@@ -1401,6 +1406,380 @@ subroutine dump_diag_yaml_obj( filename )
+          case (11)
+            call fms_f2c_string(keys3(key3_i)%key11, fileptr%file_global_meta(11,1))
+            call fms_f2c_string(vals3(key3_i)%val11, fileptr%file_global_meta(11,2))
+          case (12)
+            call fms_f2c_string(keys3(key3_i)%key12, fileptr%file_global_meta(12,1))
+            call fms_f2c_string(vals3(key3_i)%val12, fileptr%file_global_meta(12,2))
+          case (13)
+            call fms_f2c_string(keys3(key3_i)%key13, fileptr%file_global_meta(13,1))
+            call fms_f2c_string(vals3(key3_i)%val13, fileptr%file_global_meta(13,2))
+          case (14)
+            call fms_f2c_string(keys3(key3_i)%key14, fileptr%file_global_meta(14,1))
+            call fms_f2c_string(vals3(key3_i)%val14, fileptr%file_global_meta(14,2))
+          case (15)
+            call fms_f2c_string(keys3(key3_i)%key15, fileptr%file_global_meta(15,1))
+            call fms_f2c_string(vals3(key3_i)%val15, fileptr%file_global_meta(15,2))
+          case (16)
+            call fms_f2c_string(keys3(key3_i)%key16, fileptr%file_global_meta(16,1))
+            call fms_f2c_string(vals3(key3_i)%val16, fileptr%file_global_meta(16,2))
+        end select
+      enddo
+    endif
+  enddo
+  tier2size = i
+  ! gcc bug seems to occasionally ignore the start_time key when its set in the loop above
+  ! so we'll just set it again
+  do i=1, SIZE(keys2)
+    call fms_f2c_string(keys2(i)%key8, 'start_time')
+    call fms_f2c_string(vals2(i)%val8, st_vals(i))
+  enddo
+  if (DEBUG) print *, 'tier1size', 1, 'tier2size', SIZE(diag_yaml%diag_files), &
+                                                      & 'tier3size', tier3size, 'tier3each', tier3each
+  call write_yaml_from_struct_3( 'diag_out.yaml'//c_null_char,  1, keys, vals,          &
+                                 SIZE(diag_yaml%diag_files), keys2, vals2, &
+                                 tier3size, tier3each, keys3, vals3,       &
+                                 (/size(diag_yaml%diag_files), 0, 0, 0, 0, 0, 0, 0/))
+  deallocate( keys, keys2, keys3, vals, vals2, vals3)
+
+end subroutine
+
+!> private function for getting unit string from diag_data parameter values
+character(len=7) function get_diag_unit_string( unit_param )
+  integer, intent(in) :: unit_param !< A diag unit parameter value from diag_data_mod.
+                                    !! <br>eg. DIAG_SECONDS, DIAG_MINUTES,DIAG_HOURS, DIAG_DAYS, DIAG_YEARS
+  select case(unit_param)
+    case (DIAG_SECONDS)
+      get_diag_unit_string = 'seconds'
+    case (DIAG_MINUTES)
+      get_diag_unit_string = 'minutes'
+    case (DIAG_HOURS)
+      get_diag_unit_string = 'hours'
+    case (DIAG_DAYS)
+      get_diag_unit_string = 'days'
+    case (DIAG_YEARS)
+      get_diag_unit_string = 'years'
+    case default
+      get_diag_unit_string = ''
+  end select
+end function
+
+!> private function for getting reduction type string from parameter values
+character(len=7) function get_diag_reduction_string( reduction_val ) &
```

Comment:
is this pure?

Reply:
it wasn't due to mpp_error call, but i removed that and made this one and the other one pure. I moved the error checking to outside the routine